File: conductor/src/main/java/com/bluelinelabs/conductor/changehandler/SimpleSwapChangeHandler.java
package com.bluelinelabs.conductor.changehandler;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;

import com.bluelinelabs.conductor.Controller;
import com.bluelinelabs.conductor.ControllerChangeHandler;

/**
 * A {@link ControllerChangeHandler} that will instantly swap Views with no animations or transitions.
 */
public class SimpleSwapChangeHandler extends ControllerChangeHandler {

    private static final String KEY_REMOVES_FROM_ON_PUSH = "SimpleSwapChangeHandler.removesFromViewOnPush";

    private boolean removesFromViewOnPush;

    private boolean canceled;

    public SimpleSwapChangeHandler() {
        this(true);
    }

    public SimpleSwapChangeHandler(boolean removesFromViewOnPush) {
        this.removesFromViewOnPush = removesFromViewOnPush;
    }

    @Override
    public void saveToBundle(@NonNull Bundle bundle) {
        super.saveToBundle(bundle);
        bundle.putBoolean(KEY_REMOVES_FROM_ON_PUSH, removesFromViewOnPush);
    }

    @Override
    public void restoreFromBundle(@NonNull Bundle bundle) {
        super.restoreFromBundle(bundle);
        removesFromViewOnPush = bundle.getBoolean(KEY_REMOVES_FROM_ON_PUSH);
    }

    @Override
    public void onAbortPush(@NonNull ControllerChangeHandler newHandler, Controller newTop) {
        super.onAbortPush(newHandler, newTop);

        canceled = true;
    }

    @Override
    public void performChange(@NonNull ViewGroup container, View from, View to, boolean isPush, @NonNull final ControllerChangeCompletedListener changeListener) {
        if (!canceled) {
            if (from != null && (!isPush || removesFromViewOnPush)) {
                container.removeView(from);
            }

            if (to != null && to.getParent() == null) {
                container.addView(to);
            }
        }

        changeListener.onChangeCompleted();
    }

    @Override
    public boolean removesFromViewOnPush() {
        return removesFromViewOnPush;
    }
}

File: conductor/src/main/java/com/bluelinelabs/conductor/changehandler/SimpleSwapChangeHandler.kt
package com.bluelinelabs.conductor.changehandler

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import com.bluelinelabs.conductor.Controller
import com.bluelinelabs.conductor.ControllerChangeHandler

/**
 * A [ControllerChangeHandler] that will instantly swap Views with no animations or transitions.
 */
class SimpleSwapChangeHandler @JvmOverloads constructor(
  removesFromViewOnPush: Boolean = true,
) : ControllerChangeHandler(), View.OnAttachStateChangeListener {

  private var _removesFromViewOnPush = removesFromViewOnPush
  override val removesFromViewOnPush: Boolean
    get() = _removesFromViewOnPush

  override val isReusable = true

  private var canceled = false
  private var container: ViewGroup? = null
  private var changeListener: ControllerChangeCompletedListener? = null

  override fun saveToBundle(bundle: Bundle) {
    super.saveToBundle(bundle)
    bundle.putBoolean(KEY_REMOVES_FROM_ON_PUSH, removesFromViewOnPush)
  }

  override fun restoreFromBundle(bundle: Bundle) {
    super.restoreFromBundle(bundle)
    _removesFromViewOnPush = bundle.getBoolean(KEY_REMOVES_FROM_ON_PUSH)
  }

  override fun onAbortPush(newHandler: ControllerChangeHandler, newTop: Controller?) {
    super.onAbortPush(newHandler, newTop)
    canceled = true
  }

  override fun completeImmediately() {
    changeListener?.onChangeCompleted()
    changeListener = null

    container?.removeOnAttachStateChangeListener(this)
    container = null
  }

  override fun performChange(
    container: ViewGroup,
    from: View?,
    to: View?,
    isPush: Boolean,
    changeListener: ControllerChangeCompletedListener,
  ) {
    if (canceled) return

    if (from != null && (!isPush || removesFromViewOnPush)) {
      container.removeView(from)
    }
    if (to != null && to.parent == null) {
      container.addView(to)
    }

    if (container.windowToken != null) {
      changeListener.onChangeCompleted()
    } else {
      this.changeListener = changeListener
      this.container = container
      container.addOnAttachStateChangeListener(this)
    }
  }

  override fun onViewAttachedToWindow(v: View) {
    v.removeOnAttachStateChangeListener(this)

    changeListener?.onChangeCompleted()
    changeListener = null
    container?.removeOnAttachStateChangeListener(this)
    container = null
  }

  override fun onViewDetachedFromWindow(v: View) = Unit

  override fun copy(): ControllerChangeHandler = SimpleSwapChangeHandler(removesFromViewOnPush)
}

private const val KEY_REMOVES_FROM_ON_PUSH = "SimpleSwapChangeHandler.removesFromViewOnPush"
